README.md
# SimBot

SimBot is Perl IRC bot implemented via the Perl Object Environment and POE::Component::IRC. It is primarily intended as an entertainment bot and does not presently do any form of channel management. It is likely that some channel management features will become available in the future via plugins.

I don't really use IRC anymore and all the people who played with and contributed to this bot have moved on to other things, so this is just preserved for posterity here.


ChangeLog
Version 1.0 alpha: (25 Feb 2007)
	* plugins/weather.pl
	  - Added a column 'url' to the stations database. A custom station
	    can be added to the database. url must be the URL to a METAR
	    format report. Custom station IDs currently must be in the format
	    Kxxxxxxxn where x is a letter and n a number.

Version 1.0 alpha: (21 Jan 2007)
	* plugins/dice.pl:
	  - Change by KewlioMZX: When rolling dice, give the total.
	
Version 1.0 alpha: (10 Nov 2005)
	* lib/SimBot/Util.pm:
	  - Moved several functions into SimBot::Util, which can be used by
	    plugins.  Other scripts could also use this functions externally.
	  - This allows utility functions to be called without a namespace and
	    also prevents plugins from being able to access things we don't
	    want exported from the namespace.

Version 1.0 alpha: (09 Aug 2005)
	* templates/irclog.default.tmpl:
	  - Template used for making the HTML version of the IRC log. Don't like
	    the traditional "<JohnDoe> Hi!"? Change it here.
	* plugins/httpd.pl:
	  - Error pages. Plugins must return a HTTP status code on the page
	    callback. If they don't, we'll assume 500 (Internal Server Error).
	    For codes 300-500, we'll look for a error.NNN template. If we can't
	    find one, we'll use error.500. Failing that, we use a hard coded page.

Version 1.0 alpha: (08 Aug 2005)
	* plugins/httpd.pl:
	  - I forgot to update the changelog when I created this plugin...
	    instead of faking dates for everything, here it all is:
	  - Internal web server for SimBot. Right now, sqlite-logger and recap
	    use it. It also provides an admin page where you can restart SimBot
	    and even make it say things.
	  - Added templates.
	* templates/base.default.tmpl:
	  - This is the basic template that defines every page. More templates will
	    come soon...

Version 1.0 alpha: (25 Jul 2005)
	* simbot.pl:
	  - Added message value returned to callbacks attached to the nojoin
	    (banned/invite-only) event.
	  - The starting score for finding interesting words is more adaptive
	    so oft used interesting words should be able to keep a positive
	    score better.  The query_word_score callbacks now receive the
	    start score.
	  - The debug function now accepts constants in addition to numbers
	    for the different levels.  DEBUG_NO_PREFIX allows callers to avoid
	    having the PREFIX text attached to the beginning of their line.
	* plugins/services.chanserv.pl:
	  - Rewrote the chanserv support to automatically attempt to register
	    with nickserv, and to automatically try to link nicknames with
	    the nickname represented by the services "user."  Additional error
	    handling has been added as well.
	* plugins/services.dalnet.pl:
	  - Dropped in favor of handling all chanserv operations in one plugin.

Version 1.0 alpha: (23 Jul 2005)
	* plugins/8ball.pl:
	  - SimBot has acquired a Magic Eight Ball. Ask it a question!

Version 1.0 alpha: (18 Jul 2005)
	* plugins/weather.pl:
	  - Added our own METAR parser. With luck, this will be more sane.
	  - If we can figure out a valid time for the report, we tell the user
	    how old the report is instead of the UTC time.

Version 1.0 alpha: (08 May 2005)
	* plugins/time.pl:
	  - New plugin!
	  - Shows the current time in a number of time zones.
	  - %time <zones> shows the local time and the requested zones,
	    otherwise shows local and those defined in config.ini.
	    Failing that, local and UTC.

Version 1.0 alpha: (07 May 2005)
	* plugins/weather.pl:
	  - Some unit conversion support. Add 'm' or 'metric' to show metric
	    units, 'us' or 'imp' or 'imperial' for imperial units.
	  - Support for UK post codes
	* tools/create_postalcode_db.pl:
	  - Replaces tools/create_zip_db.pl
	  - Includes UK post codes

Version 1.0 alpha: (06 May 2005)
	* plugins/weather.pl:
	  - Support for %weather <zip> to look up the closest station for
	    current observations.

Version 1.0 alpha: (05 May 2005)
	* tools/create_zip_db.pl:
	  - Creates the USzip file, a SQLite DB of US zip codes circa 1999 from
	    US census data. This file lets us match zip codes to lat/long.
	* plugins/weather.pl:
	  - Slightly better forecast support, you can now use
	    %weather forecast <zip> to get the forecast for your zip code.

Version 1.0 alpha: (04 May 2005)
	* plugins/weather.pl:
	  - Forecast support; %weather <lat> <long> will give a forecast or break
	    horribly depending on how sober NOAA is and if the lat/long is in the
	    US.
	* simbot.pl:
	  - Replace plugin_desc parameter with plugin_help and plugin_params
	    to allow a very nicely formed usage description.  Plugins that have
	    no help will not be listed, even if they have params, but if the
	    user requests help on them, the params will be revealed if they
	    exist.
	* plugins/sqlite-logger.pl
	  - The asterix (*) can now be specified for the nickname, this will look
	    for anybody matching. This replaces %log last.

Version 1.0 alpha: (03 May 2005)
	* plugins/seen.pl:
	  - Removed, replaced with sqlite-logger.pl's seen support
	* plugins/sqlite-logger.pl
	  - Now supports %seen. %log seen will hang around for the time being.
	  - %seen <nick> <events> count <number> will give multiple matches, up
	    to <number>.
	  - %seen <nick> <events> content <words> will look in the log for the
	    phrase <words>.
	  - Gives a word bonus to nicknames, like seen.pl used to.
	* simbot.pl:
	  - Now properly stops sending data to the server when trying to quit.

Version 1.0 alpha: (01 May 2005)
	* plugins/currency.pl:
	  - Rewritten! Now uses SOAP::Lite to fetch exchange information instead
	    of stealing it from xe.com.

Version 1.0 alpha: (30 Apr 2005)
	* plugins/aspell.pl:
	  - New plugin! Checks and offers suggestions for the user's spelling.
	    Requires Text::Aspell.

Version 1.0 alpha: (31 Dec 2004)
	* Happy New Year!
	* plugins/sqlite-logger.pl:
	  - Added a %log last <event> command, which will tell you when

plugins/dice.pl
###
#  SimBot Dice & Coin Plugin
#
# DESCRIPTION:
#   Every IRC needs to be able to roll some dice of arbitrary sides for its
#   regulars, and SimBot is no exception. Responds to '%roll' with a pair of
#   6 sided dice, or to '%roll xdy' with x y-sided dice. It can also flip a
#   coin with '%flip' and play "rock, paper, scissors, with '%rps.'
#
# COPYRIGHT:
#   Copyright (C) 2003-07, Pete Pearson
#
#   This program is free software; you can redistribute and/or modify it
#   under the terms of version 2 of the GNU General Public License as
#   published by the Free Software Foundation.
#
#   This program is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#   GNU General Public License for more details.
#   
#   You should have received a copy of the GNU General Public License
#   along with this program; if not, write to the Free Software
#   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
#
# TODO:
#   *
#

package SimBot::plugin::roll;

use strict;
use warnings;

# Use the SimBot Util perl module
use SimBot::Util;

# Jeers for Rock Paper Scissors
use constant RPS_JEER => (
						  'Oh, the horror!',
						  'Your luck won\'t hold up forever!',
						  'I\'ll beat you next time.',
						  'Damn!',
						  'Don\'t let the success go to your head!',
						  'I will reign victorious... later.',
						  );

# Cheers for Rock Paper Scissors
use constant RPS_CHEER => (
						   'In your face!',
						   'I\'m on a roll.',
						   'Behold my awesome power!',
						   'Victory is mine!',
						   'I am the RPS grand master!',
						   'You shall not defeat me!',
						   );

sub roll_dice {
    my $numDice = 2;
    my $numSides = 6;
    my $diceTotal = 0;
    my ($kernel, $nick, $channel, undef, $dice) = @_;
    if($dice =~ m/(\d*)[Dd](\d+)/) {
        $numDice = (defined $1 ? $1 : 1);
        $numSides = $2;
    }
    if($numDice == 0) {
        &SimBot::send_message($channel, "$nick: I can't roll zero dice!");
	} elsif($numDice > 100000000000000) {
        &SimBot::send_message($channel, "$nick: I can't even count that high!");
    } elsif($numDice > 100) {
        &SimBot::send_action($channel, "drops $numDice ${numSides}-sided dice on the floor, trying to roll them for ${nick}.");
    } elsif($numSides == 0) {
        &SimBot::send_action($channel, "rolls $numDice zero-sided " . (($numDice==1) ? 'die' : 'dice') . " for ${nick}: " . (($numDice==1) ? "it doesn't" : "they don't") . ' land, having no sides to land on.');
    } elsif($numSides > 1000) {
        &SimBot::send_message($channel, "$nick: The numbers on the dice are so small that I can't read them!");
    } else {
        my @rolls = ();
        for(my $x=0;$x<$numDice;$x++) {
	    my $diceRoll = int rand($numSides)+1;
	    $diceTotal += $diceRoll;
            push(@rolls, $diceRoll);
        }

        &SimBot::send_action($channel, "rolls $numDice ${numSides}-sided " . (($numDice==1) ? 'die' : 'dice') . " for ${nick}: " . join(' ', @rolls) . ($numDice==1 ? '' : " -- total of $diceTotal"));
    }
}

sub flip_coin {
    my ($kernel, $nick, $channel, undef, $number) = @_;
	my $text;
	$number = 1 if !defined $number;
	if ($number > 20) {
		&SimBot::send_message($channel, "$nick: It's dangerous to throw that many coins in the air at once.");
	} elsif ($number <= 0) {
		&SimBot::send_action($channel, "tries to flip $number coins for $nick, but the universe appears firmly opposed to the idea.");
	} else {
		$text = "";
		for (my $i=1; $i < $number; $i++) {
			$text .= ((int rand(2)==0) ? 'heads, ' : 'tails, ');
		}
		$text .= ((int rand(2)==0) ? 'heads.' : 'tails.');
		&SimBot::send_action($channel, "flips " .
							 ($number > 1 ? "$number coins" : "a coin") .
							 " for $nick: " . $text);
	}
}

sub rps_shoot {
    my ($kernel, $nick, $channel, undef, @words) = @_;
	my $object = lc(join(" ", @words));
	my $text;
	if (!defined $object || $object eq "") {
		&SimBot::send_message($channel, "$nick: Well?  Rock, paper, or scissors?");
	} elsif ($object =~ /^(rock|paper|scissors)$/) {
		my $pick = &pick(('rock', 'paper','scissors'));
		$text = "$nick: I picked: $pick. ";
		if ($pick eq $object) {
			$text .= "It's a draw! Try again!";
		} elsif ($object eq "rock") {
			if ($pick eq "scissors") {
				$text .= "Your rock smashed my scissors. " . &pick(RPS_JEER) . " You win!";
			} elsif ($pick eq "paper") {
				$text .= "My paper covered your rock. " . &pick(RPS_CHEER) . " I win!";
			}
		} elsif ($object eq "paper") {
			if ($pick eq "rock") {
				$text .= "Your paper covered my rock. " . &pick((RPS_JEER)) . " You win!";
			} elsif ($pick eq "scissors") {
				$text .= "My scissors cut your paper. " . &pick((RPS_CHEER)) . " I win!";
			}
		} elsif ($object eq "scissors") {
			if ($pick eq "paper") {
				$text .= "Your scissors cut my paper. " . &pick((RPS_JEER)) . " You win!";
			} elsif ($pick eq "rock") {
				$text .= "My rock smashed your scissors. " . &pick((RPS_CHEER)) . " I win!";
			}
		}
		&SimBot::send_message($channel, $text);
	} else {
		&SimBot::send_message($channel, "$nick: You have been disqualified for bringing your $object to a rock, paper, scissors game!  You forfeit!  I win!");
	}
}

sub nlp_match {
    my ($kernel, $nick, $channel, $plugin, @params) = @_;
	my $sides = 6;
	my $dice = 2;
	my $coins = 1;
	my %quantities = (
					  "some", "4",
					  "few", "3",
					  "couple", "2",
					  "couple of", "2",
					  "handful of", "10",
					  "a", "1",
					  "no", "0",
					  "an", "1",
					  "the", "1",
					  "many", "12",
					  "dozen", "12",
					  "dozens of", "48",
					  "several", "7",
					  "lot of", "30",
					  "lots of", "30",
					  "copious", "42",
					  "plenty of", "72",
					  "excessive", "1000",
					  "every", "100000000000001",
					  "all the", "100000000000001",
					  "all", "100000000000001",
					  "0", "0",
					  );
	my $qmatch = "(" . join("|", keys(%quantities)) . ")";
	$qmatch = qr/$qmatch/i;

	if ($plugin eq "roll") {
		foreach (@params) {
			no warnings;
			if (m/(\d+|$qmatch) (die|dice)/i) {
				$dice = ($1 != 0 ? $1 : $quantities{$1});
			} elsif (m/\bdie\b/i) {
				$dice = 1;
			}

			if (m/(\d+|$qmatch) (\d+|$qmatch)-sided/i) {
				$dice = ($1 != 0 ? $1 : $quantities{$1});
				$sides = ($2 != 0 ? $2 : $quantities{$2});
			} elsif (m/(\d+|$qmatch)-sided/i) {
				$sides = ($1 != 0 ? $1 : $quantities{$1});
			} elsif (m/(\d+|$qmatch) (side|sides|sided)/i) {
				$sides = ($1 != 0 ? $1 : $quantities{$1});
			} elsif (m/\bside\b/i) {
				$sides = 1;
			}
		}

		&roll_dice($kernel, $nick, $channel, undef, $dice . "d" . $sides);
	}

	if ($plugin eq "flip") {
		foreach (@params) {
			no warnings;
			if (m/(\d+|$qmatch) (coin|coins)/i) {
				$coins = ($1 != 0 ? $1 : $quantities{$1});
			} elsif (m/(coins)/i) {
				$coins = 2;
			}
		}

		&flip_coin($kernel, $nick, $channel, undef, $coins);
	}

	return 1;
}

# Register Plugins
&SimBot::plugin_register(plugin_id   => "roll",
						 plugin_help => "Rolls dice. You can specify how many dice, and how many sides, in the format 3D6.",
						 plugin_params => "[<dice>[D<sides>]]",

						 event_plugin_call => \&roll_dice,

						 event_plugin_nlp_call => \&nlp_match,
						 hash_plugin_nlp_verbs =>
						 ["roll", "throw"],
						 hash_plugin_nlp_subjects =>
						 ["dice", "die"],
						 hash_plugin_nlp_formats =>
						 ["{q} die", "{q} dice",
						  "{has} {q} sides", "{has} {q} side",
						  "{q} {q}-sided", "{q}-sided",
						  "{q} {q} sided", "{q} sided", ],
						 hash_plugin_nlp_questions =>
						 ["you-must", "you-should", "you-may",
						  "what-if", "how-about", "i-need",
						  "i-want", "would-you", "command",],
						 );

&SimBot::plugin_register(plugin_id   => "flip",
						 plugin_params => "[<number of coins>]",
						 plugin_help => "Flips a coin.",

						 event_plugin_call => \&flip_coin,

						 event_plugin_nlp_call => \&nlp_match,
						 hash_plugin_nlp_verbs =>
						 ["flip", "toss"],
						 hash_plugin_nlp_subjects =>
						 ["coin", "coins"],
						 hash_plugin_nlp_formats =>
						 ["{q} coins", "{q} coin", ],
						 hash_plugin_nlp_questions =>
						 ["you-must", "you-should", "you-may",
						  "what-if", "how-about", "i-need",
						  "i-want", "would-you", "command",],
						 );

&SimBot::plugin_register(plugin_id   => "rps",
						 plugin_params => "<selection>",
						 plugin_help => "Pick rock, paper or scissors.",
						 event_plugin_call => \&rps_shoot,
						 );
